File: joss/comparisons/cat-3-null-missing-data/21-aggregation-return-type/example-aggregation-return-type.R
# Error Class 21: Aggregation Return Type Narrowing
#
# R's aggregation functions return NA when any value is NA (without na.rm).
# With na.rm=TRUE, they return a clean numeric. But there's no type-level
# difference — both are "numeric". You can do arithmetic on NA results
# and it just propagates NA silently.

library(tidyverse)

labs <- tibble(
  patient_id = c("P001", "P001", "P002"),
  test_name = c("BNP", "WBC", "BNP"),
  result_value = c(1250, NA, 450)
)

# SILENT: sum() returns NA — no warning
total <- sum(labs$result_value)
print(total)  # NA

# SILENT: Arithmetic on NA propagates — no warning
per_patient <- total / 2  # NA
print(per_patient)  # NA — wrong result, no indication

File: joss/comparisons/cat-3-null-missing-data/21-aggregation-return-type/example-aggregation-return-type.py
"""
Error Class 21: Aggregation Return Type Narrowing

Python's aggregation functions silently skip NaN values by default.
The return type is always float64 — there's no type-level distinction
between "sum with NaN removed" and "sum of clean data".
"""
import pandas as pd
import numpy as np

labs = pd.DataFrame({
    "patient_id": ["P001", "P001", "P002"],
    "test_name": ["BNP", "WBC", "BNP"],
    "result_value": [1250, np.nan, 450],
})

# SILENT: sum() skips NaN by default — no indication that data was incomplete
total = labs["result_value"].sum()  # 1700.0, not NaN
print(f"sum = {total}")  # Looks correct but silently excluded missing data

# SILENT: No type difference between clean sum and NaN-skipping sum
per_patient = total / 2  # Works fine — but was the NaN intentional?
print(f"per_patient = {per_patient}")
